### ISSUE TITLE
[java] CloneMethodMustImplementCloneable: False positive when Cloneable is a super interface

### ISSUE THREAD

**Issue by mebigfatguy** [pmd/pmd#6345]
**Affects PMD Version:** 7.19.0

**Rule:** [CloneMethodMustImplementCloneable](https://docs.pmd-code.org/latest/pmd_rules_java_errorprone.html#clonemethodmustimplementcloneable)

**Description:**

if a class implements the clone method and implements an interface, and that interface has as a super interface Cloneable, then PMD reports it has failing this rule, when in fact it does inherit from Cloneable

**Code Sample demonstrating the issue:**

```java
public interface MyC extends Cloneable {
}

public class MyExample implements MyC {
    public MyExample clone() {
           try {
                 return (MyExample) super.clone();
           } catch (CloneNotSupportedException e) {
                 throw new Error();
           }
    }
}
```

**Expected outcome:**

PMD reports a violation at line ..., but that's wrong. That's a false positive.

**Running PMD through:** *[CLI | Ant | Maven | Gradle | Designer | Other]*
ant

**Comment by mdhamed238** [comment]
I'm on it.

**Comment by mdhamed238** [comment]
Hi @mebigfatguy,

I have verified this code sample against PMD 7.20.0 and can confirm that the issue is no longer reproducible. The rule now correctly identifies the usage, and no false positive is reported.

I recommend upgrading to the latest version to resolve this.

**Comment by pzygielo** [comment]
> I recommend upgrading to the latest version to resolve this.

Just to confirm - at this very moment 7.19.0 IS the latest version, right?

**Comment by pzygielo** [comment]
> I have verified this code sample against PMD 7.20.0 and can confirm that the issue is no longer reproducible. The rule now correctly identifies the usage, and no false positive is reported.

Is it possible to clarify which code change in https://github.com/pmd/pmd/releases/tag/pmd_releases%2F7.20.0-SNAPSHOT changed/fixed that?

**Comment by mdhamed238** [comment]
Hi @pzygielo, sorry for the inconvenience. I downloaded the official release 7.20.0, and this appears to be a confirmed false positive.  

```bash
pmd check \
  -d MyExample.java \
  -R category/java/errorprone.xml,category/java/bestpractices.xml \
  -f text

---

[WARN] Progressbar rendering conflicts with reporting to STDOUT. No progressbar will be shown. Try running with argument -r <file> to output the report to a file instead.
[WARN] This analysis could be faster, please consider using Incremental Analysis: https://docs.pmd-code.org/latest/pmd_userdocs_incremental_analysis.html
MyExample.java:5:        MissingOverride:    The method 'clone()' is missing an @Override annotation.
MyExample.java:9:    PreserveStackTrace:  Thrown exception does not preserve the stack trace of exception 'e' on all code paths
```

It turns out that I didn’t reproduce the issue correctly locally. I’ll get back to you as soon as I do

**Comment by mdhamed238** [comment]
Update: The bug occurs when the interface and class are in separate files.

**MyC.java:**
```java
public interface MyC extends Cloneable {
}
```

**MyExample.java:**
```java
public class MyExample implements MyC {
    public MyExample clone() {
        try {
            return (MyExample) super.clone();
        } catch (CloneNotSupportedException e) {
            throw new Error();
        }
    }
}
```

**Output:**
```
MyExample.java:3: CloneMethodMustImplementCloneable: clone() method should be implemented only if implementing Cloneable interface
```

This is a **false positive**. The rule fails to recognize that `MyExample` implements `Cloneable` transitively through `MyC` when they're defined in separate files.

My initial test had both declarations in a single file, which is why I couldn't reproduce it originally. The issue is **confirmed in PMD 7.20.0**.

@pzygielo - Is checking interface hierarchies across separate files the expected behavior for this rule, or should it require auxclasspath configuration?

**Comment by jsotuyod** [comment]
@mdhamed238 are you sure you are providing a complete `auxclasspath`? the fact it works when using a single source files makes me believe you are not…

Without a complete auxclasspath, PMD simply can't know what is in `MyC`, or wether it implements `Cloneable` or not.

The original report by @mebigfatguy mentions he is using `Ant`, which does support the param, but requires manual setting of it + ensuring the code is built before running PMD (unlike Maven and Gradle that manages everything automatically).

**Comment by mdhamed238** [comment]
@jsotuyod Thanks for the clarification. So --aux-classpath is needed to tell PMD where the compiled classes are, allowing it to resolve the interface hierarchy (MyC → Cloneable) across separate files.

**Comment by UncleOwen** [comment]
This exact feature has been implemented in [#1532 on sourceforge](https://sourceforge.net/p/pmd/bugs/1532/) and has tests that verify it works.

I just tried the exact code from the first post in current main and didn't get a warning. Closing as can't reproduce.
